### PR TITLE
Add new tables for versioning and changes, respective CRUD stuff

### PR DESCRIPTION
**Pull request by dakota002**
Backend breakout for further updates with versioning

### CODE REVIEW

File: app/routes/circulars/circulars.lib.ts
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:129]: Move these two keyword additions to a separate PR.

**Review by lpsinger — CHANGES_REQUESTED** [review]
- Move circular keywords into a separate table.

File: app/routes/circulars/circulars.lib.ts
Comment (lpsinger) [reply app/routes/circulars/circulars.lib.ts:129]: I don't understand. Do people request changes to a Circular by submitting a Circular with these magic works in the subject?

Comment (dakota002) [reply app/routes/circulars/circulars.lib.ts:129]: Ah I see, no I think I am seeing an error in my implementation. I don't want the users to submit circulars where the value is literally 'Duplicate Submission', I was checking that the string was valid, but it shouldn't be the case that these are just considered valid keywords. 

In the [big PR](https://github.com/nasa-gcn/gcn.nasa.gov/pull/1425) I had a couple simplified workflows that would be shortcuts for mods to just click a button and it would handle the workflow for duplicates/mistaken submissions

I have some ideas for how to improve this, the functionality as I added it is wrong, I'll update this shortly

**Review by lpsinger — CHANGES_REQUESTED** [review]
- What is the purpose of `erroneousSubmissionKeywords`? Do you want to require users to insert one of these keywords into the subject lines for proposed edits of GCN Circulars? Why? The fact that a Circular has a version number that is greater than one will indicate that it has been edited, and we can use that to mark it differently in the UI.
- We should be careful to generate new versions using transactions to avoid race conditions. Fortunately, the logic should be almost identical to what we do in dynamodb-autoincrement. Would you please propose a patch for dynamodb-autoincrement to either (1) generalize it to support version control applications or (2) add a second class that implements version control?

File: app/routes/circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/circulars/circulars.server.ts:339]: Not if the subject contains an `erroneousSubmissionKeyword`, but that `erroneousSubmissionKeywords` includes the subject.

If the subject is not valid and the subject is not literally one of the `erroneousSubmissionKeywords`: throw error

so if either the subject is valid, or the subject is one of the listed items in the `erroneousSubmissionKeywords`, it will continue, no?

Comment (lpsinger) [reply app/routes/circulars/circulars.server.ts:339]: Oh, I get it now.

File: app/routes/circulars/circulars.lib.ts
Comment (lpsinger) [inline app/routes/circulars/circulars.lib.ts:171]: Rename to make it clear that these are not partial matches, these are whole subjects.

```suggestion
export const erroneousSubmissionSubjects = [
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/circulars/circulars.lib.ts
Comment (dakota002) [reply app/routes/circulars/circulars.lib.ts:171]: Good idea, I'll have that in the next commit

**Comment by dakota002** [comment]
Converting to draft PR, need to revise the implementation further

**Comment by lpsinger** [comment]
You've requested a review from me, but this is still marked as a draft. Is it ready for review? If so, then please remove the draft status.

**Comment by lpsinger** [comment]
> Converting to draft PR, need to revise the implementation further

Oh, OK. Let me know when you are ready for a review.

**Comment by dakota002** [comment]
Okay, should be ready now. Had to strip out the stuff that the auto-incrementer was updated to handle

**Comment by dakota002** [comment]
Rebased, should resolve the conflicts

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:77]: This will store a memoized autoincrement object _for each circular ID_. That's too much!

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:340]:
```suggestion
  return result.Items
```

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:323]: Let's merge this functionality into the existing `get()` function, adding an optional version positional argument.

Also: since the _current_ version is not stored in the history table, if the caller requests the current version by its number, then we need to correctly handle that case.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:330]: - No need for "Circular" to be in the name... it's clear because this is in circulars.server.ts.
- Consider `getVersions()` instead of `getHistory()` to reduce the number of different similar but distinct terms in function names (history, version, change)
- Is it more useful to return the content of all versions, or simply a list of their version numbers? Perhaps version numbers and titles, like `CircularMetadata[]`? I'm thinking that the main use of this function will be to implement an index route that lists all of the versions.

```suggestion
export async function getVersions(
```

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:372]: This duplicates much of the logic from the existing `put` function. Please refactor.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:400]: Again, do we want this to return the entire content of all of the change requests, or just a list of them?

File: app.arc
Comment (lpsinger) [inline app.arc:83]: Will this data model limit us to one change request per circular per user at a time?

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:371]: Should we only permit users who are submitters to enter change requests?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:371]: I don't know. @jracusin?

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:77]: Please _don't_ memoize this.

**Review by lpsinger — CHANGES_REQUESTED** [review]
I see that you requested another review from me, but there are still unresolved comments from the last review.

File: app.arc
Comment (dakota002) [reply app.arc:83]: Yes. I think that makes the most sense for this use case. Once approved, this just sets the subject and body to whatever the request has in it, so if they make multiple requests at the same they will result in overwriting. 

If we see more of a need for some granularity in the future we could implement more of a diff based approach

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts]: Yup! updated in latest commit

**Comment by dakota002** [comment]
> I see that you requested another review from me, but there are still unresolved comments from the last review.

Oh that shouldn't have triggered the re-review until now, I was still working on a commit. It should be ready for review now though, with the exception of the point waiting for @jracusin's feedback

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:227]: A couple issues with this:

- The code within the `if (version) { if (isNaN(version)) { ... }}` block cannot be reached because `NaN` is falsey.
- When getting a Circular by version, we _still_ need to handle the case that the current version _is_ the requested version.
- The new `if (!result)` section is redundant because that check already exists below.

One possible refactoring:

```suggestion
  if (version) {
    result = await db.circulars_history.get({
      circularId,
      version,
    })
  }
  if (!result && !isNaN(version)) {
    result = await db.circulars.get({
      circularId
    })
  }
  if (version && result?.version !== version)
  {
    result = undefined
  }
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:322]: Do we want to return the _content_ of all of the circulars, or just a list of their version numbers?

Also, this does not include the newest or current version number.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:399]:
```suggestion
  if (!user) throw new Response(null, { status: 403 })
```

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:393]: None of the other exported functions have names that start with `verifyAnd`. Suggest renaming this `verifyAndDeleteChangeRequest` and the internal function `deleteChangeRequestInternal` or `deleteChangeRequestRaw`.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:454]: Factor out as `getChangeRequest`?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:322]: It should include all versions that exist, since the auto increment puts into both tables, right? I'll need to test a bit. I think in the previous iteration I assumed that the history didn't contain the latest, and I just joined the history and the current entry in the circulars table. I can edit this to make sure it returns the whole list

Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:322]: Remember that in @nasa-gcn/dynamodb-autoincrement we landed on a design in which all versions _except for the most recent one_ are stored in the history table.

I suggest the following:

- Change the signature of this function so that it returns a list of version numbers.
- Add a projection to the query so that it only returns the circularId attribute.
- Before returning, sort the versions, and push one additional implied version.

Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts]: Ah right, okay I updated this and another bug in the get method in this latest commit

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:220]: Remember that `Boolean(NaN) === false`. This `if` statement is not correct.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:327]: Note that this will return `[1]` if (a) the Circular exists but has only one version or (b) the Circular does not exist. Do we care to be correct for the latter corner case?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:220]: ah good catch, I think I am misunderstanding what this case should be catching. Is this correct:

If the result isn't found in the initial check given a `circularId` and `version` (which is a number), just get whatever the latest/current version is. 

Then in the following block, if the provided version doesn't match on the found result, clear the result

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:327]: I think the latter shouldn't be that much of an issue, since it is just returning an array, it won't really mean anything until we put it into further context, like generating links to view the versions

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:220]: Correct.

The following possible values of type `number | undefined` are falsey:

- undefined: no version passed as an argument, so don't even bother looking up by version
- NaN: an impossible value, should always throw 404
- 0: another impossible value, should always throw 404

Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:220]: Gotcha, I had made the change to that if statement because of this error:
image

Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:220]: To satisfy type inference, you chould check that `number !== undefined && !number`.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:228]: So, the first ternary expression here still isn't quite right if we want a value of `undefined` to trigger lookup of the latest version but `0` or `NaN` to trigger an error.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:228]: I see, do we just want the checks to short out in those cases? So move the check for If `0` or `NaN` ... up before doing any gets?

Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:228]: You could... or you could structure the body of the function like this.

```ts
if (version === undefined) {
  // get most recent version
} (else if !version) {
  // throw 404
} else {
  // look up by version number
}
```

Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:228]: Ah okay I didn't see this comment until after I pushed, what do you think of the latest refactor? It moves the errors up to handle it all at the beginning

Comment (lpsinger) [reply app/routes/_gcn.circulars/circulars.server.ts:228]: Ugh. You know what, if `version === 0` or `version === NaN`, I don't think you need to do anything special, since it will never match a real circular in the table.

Maybe this would be simpler using [function overloading](https://dev.to/zirkelc/how-to-overload-functions-in-typescript-22c3)?

```ts
export async function get(circularId: number) {
  ...
}

export async function get(circularId: number, version: number) {
  ...
}
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by dakota002** [comment]
True, maybe we can just scrap some of those lines? 

As for the overloading, I don't think that will help in this case. It looks like the way that TS supports overload methods is that you just define all the overloads at the top, then account for all cases within the function. So it would look like this (i removed some unneeded curly brackets to shrink it for this code block):

```ts
export async function get(circularId: number): Promise<Circular>
export async function get(
  circularId: number,
  version?: number
): Promise<Circular> {
  if (isNaN(circularId)) throw new Response(null, { status: 404 })
  if (version === 0 || (version !== undefined && isNaN(version)))
    throw new Response(null, { status: 404 })

  const db = await tables()
  let result = version
    ? await db.circulars_history.get({ circularId, version })
    : await db.circulars.get({ circularId })

  if (!result && version)
    result = await db.circulars.get({
      circularId,
    })

  if (version && result?.version !== version) result = undefined

  if (!result)
    throw new Response(null, {
      status: 404,
    })
  return result
}
```
 
Attempting to define 2 distinct methods with the same name throws these errors:

image

**Comment by lpsinger** [comment]
> True, maybe we can just scrap some of those lines?

I'll make a PR to propose a version of this function. One moment.

**Comment by lpsinger** [comment]
@dakota002, I tried myself to write list and get methods that include the current version, and I found corner cases that were difficult to correctly deal with performantly. I think that your original approach of excluding the current version from the list is better.

I opened https://github.com/nasa-gcn/dynamodb-autoincrement/pull/39 to implement this in the library.

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:204]: You can make this function a lot simpler now:

```
/** Get a circular by ID. */
export async function get(circularId: number, version?: number) {
  const circularVersions = await getDynamoDBVersionAutoIncrement(circularId)
  const result: Circular = await circularVersions.get(version)
  if (!result)
    throw new Response(null, {
      status: 404,
    })
  return result
}
```

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:309]:
```suggestion
export async function getVersions(circularId: number) {
  const circularVersionsAutoIncrement =
    await getDynamoDBVersionAutoIncrement(circularId)
  return await circularVersionsAutoIncrement.list()
}
```

File: package.json
Comment (lpsinger) [inline package.json:31]: Please rebase from the main branch. This change is already there.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (dakota002) [reply app/routes/_gcn.circulars/circulars.server.ts:204]: Sweet, updated in the latest commit with a rebase

File: app/routes/_gcn.circulars/circulars.server.ts
Comment (lpsinger) [inline app/routes/_gcn.circulars/circulars.server.ts:378]:
```suggestion
  if (requestorSub !== user.sub || !user.groups.includes(moderatorGroup))
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Pull request merged**
